src/array/cpu/concurrent_id_hash_map.h

```diff
@@ -0,0 +1,205 @@
+/**
+ *  Copyright (c) 2019 by Contributors
+ * @file dgl/concurrent_id_hash_map.h
+ * @brief Utility classes and functions for DGL arrays.
+ */
+
+#ifndef DGL_ARRAY_CONCURRENT_HASH_MAP_H_
+#define DGL_ARRAY_CONCURRENT_HASH_MAP_H_
+
+#include <vector>
+#include <atomic>
+#include <omp.h>
+#include <algorithm>
+#include <dgl/aten/types.h>
+
+#define PICK_THREAD_NUM 16
+
+namespace dgl {
+namespace aten {
+
+template <typename IdType>
+class ConcurrentIdHashMap {
+ public:
+    struct Mapping {
+        std::atomic<IdType> key;
+        IdType value;
+    };
+
+    ConcurrentIdHashMap() : hmap(NULL) {}
+
+    void Init(size_t n) {
+        size_t new_size;
+        for (new_size = 1; new_size < 3 * n; new_size *= 2) {}
+        mask =  static_cast<IdType>(new_size - 1);
+        
+        DGLContext ctx;
+        ctx.device_type = kDGLCPU;
+        auto device = runtime::DeviceAPI::Get(ctx);
+        hmap = static_cast<Mapping*>(
+            device->AllocWorkspace(ctx, sizeof(Mapping) * new_size));
+
+        // parallel memset can accelerate it
+        memset((void*)hmap, -1, sizeof(Mapping)*new_size);
+    }
+
+    size_t Update(IdArray ids,  IdArray& unique_ids) {
```

Comment:
Is `Update` only allowed to be called once?  If so, we could probably merge this into `Init` (or even better into the ctor)?

Reply:
Make sense, will move it into Init.